### PR TITLE
implement users store and access control

### PR DESCRIPTION
**Pull request by javeme**
implement: #209
Change-Id: I6492c284b6677b9dd1ef545aaf08c5d6e5cf8ad7

### CODE REVIEW

**Comment by javeme** [comment]
rebased on v0.11 and close #704

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeBelong.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        public static final String LABEL = T.label.getAccessor();
+
+        public static final String USER = HugeUser.P.USER;
+        public static final String GROUP = HugeGroup.P.GROUP;
+
+        public static final String DESCRIPTION = "~belong_description";
+        public static final String CREATE = "~belong_create";
+        public static final String UPDATE = "~belong_update";
+
+        public static String unhide(String key) {
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeBelong.java]: unHide

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeGroup.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            map.put(Hidden.unHide(P.DESCRIPTION), this.description);
+        }
+        map.put(Hidden.unHide(P.CREATE), this.create);
+        map.put(Hidden.unHide(P.UPDATE), this.update);
+
+        return map;
+    }
+
+    public static HugeGroup fromVertex(Vertex vertex) {
+        HugeGroup entity = new HugeGroup((Id) vertex.id());
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeGroup.java]: rename to group

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeTarget.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        map.put(Hidden.unHide(P.URL), this.url);
+
+        map.put(Hidden.unHide(P.CREATE), this.create);
+        map.put(Hidden.unHide(P.UPDATE), this.update);
+
+        return map;
+    }
+
+    public static HugeTarget fromVertex(Vertex vertex) {
+        HugeTarget entity = new HugeTarget((Id) vertex.id());
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeTarget.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeUser.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+import com.baidu.hugegraph.auth.SchemaDefine.Entity;
+import com.baidu.hugegraph.backend.id.Id;
+import com.baidu.hugegraph.schema.VertexLabel;
+import com.baidu.hugegraph.type.define.DataType;
+import com.baidu.hugegraph.util.E;
+
+public class HugeUser extends Entity {
+
+    /*
+     * TODO: add vertex label: group and action (or group and graph)
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeUser.java]: delete

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeUser.java [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        }
+
+        map.put(Hidden.unHide(P.CREATE), this.create);
+        map.put(Hidden.unHide(P.UPDATE), this.update);
+
+        return map;
+    }
+
+    public static HugeUser fromVertex(Vertex vertex) {
+        HugeUser entity = new HugeUser((Id) vertex.id());
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeUser.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/SchemaDefine.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+        public abstract Map<String, Object> asMap();
+
+        protected abstract void property(String key, Object value);
+        protected abstract Object[] asArray();
+
+        public static <T extends Relationship> T fromEdge(Edge edge,
+                                                          T relationship) {
+            E.checkArgument(edge.label().equals(relationship.label()),
+                            "Illegal edge label '%s' for entity '%s'",
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/SchemaDefine.java]: "Illegal edge label '%s' for relationship '%s'",

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeBelong.java (hunk 1 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/auth/HugeBelong.java]: seems there is unhide word

File: hugegraph-core/src/main/java/com/baidu/hugegraph/util/StringEncoding.java [hunk 7]
```diff
@@ -117,10 +132,14 @@ public static String decompress(byte[] value) {
         }
     }
 
+    public static String sha256(String string) {
+        byte[] stringBytes = encode(string);
+        DIGEST.reset();
+        return Base64.getEncoder().encodeToString(DIGEST.digest(stringBytes));
+    }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/util/StringEncoding.java]: BCrypt.hashpw(myPassword, BCrypt.gensalt(4))
BCrypt.checkpw(candidatePassword, dbPassword)

**Pull request merged**